ethers-middleware/src/proxy_wallet/ds_proxy/mod.rs

```diff
@@ -0,0 +1,89 @@
+use super::{ProxyWallet, ProxyWalletError};
+use ethers_contract::BaseContract;
+use ethers_core::{
+    abi::{parse_abi, Tokenize},
+    types::*,
+};
+
+const DS_PROXY_EXECUTE_SIGNATURE: &str =
+    "function execute(address target, bytes data) public payable returns (bytes memory response)";
+
+#[derive(Debug)]
+pub struct DsProxy {
+    address: Address,
+}
+
+impl DsProxy {
+    /// Create a new instance of DsProxy by providing the address of the DsProxy contract that has
+    /// already been deployed to the Ethereum network.
+    pub fn new(address: Address) -> Self {
+        Self { address }
+    }
+}
+
+impl ProxyWallet for DsProxy {
+    fn get_proxy_tx(&self, tx: TransactionRequest) -> Result<TransactionRequest, ProxyWalletError> {
+        // clone the tx into a new proxy tx.
+        let mut proxy_tx = tx.clone();
+
+        // the target address cannot be None.
+        let mut target = Address::default();
+        if let Some(NameOrAddress::Address(addr)) = tx.to {
+            target = addr;
+        } else {
+            return Err(ProxyWalletError::Dummy);
+        }
+
+        // fetch the data field.
+        let data = tx.data.unwrap_or(vec![].into());
+
+        // encode data as the ABI encoded data for DSProxy's execute method.
+        let ds_proxy_base: BaseContract = parse_abi(&[DS_PROXY_EXECUTE_SIGNATURE])?.into();
```

Comment:
Probably better for us to move this `BaseContract` in the DsProxy constructor, instead of re-parsing the ABI each time.